A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+  }
+
+  repeated Route route = 1;
+  map<string, Action> action = 2;
+}
+```
+
+The route refers to the action name instead of embedding the action directly,
+so that one action can be shared by multiple routes.
+
+The config must fulfill the following:
+
+*   Routes
+    *   There must be exactly one `path_matcher` rule.
+        *   `path_matcher` is an oneof so this is always true in proto. But json
+            doesn’t have the oneof concept, so a route may have multiple path
+            matchers, and it will be considered invalid.
+    *   The `action` name must be present in the action list.
+    *   Duplicate matchers are allowed.
+        *   Duplicate matcher means that more than one routes have the same
+            matcher, no matter whether the actions are the same or not.
+        *   Because the first match wins, the second matcher will never be used.
+*   Actions
+    *   The actions in list must be referred by the routes
+    *   There must be at least one valid config in `child_policy`
+
+Violating any of these will make the balancer config invalid, and be rejected.
+
+For each action, the config in `child_policy` to be used to create a child LB
+policy will be picked using the same algorithm as [service
+config](https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L297).
+
+*   The first known (registered) policy name in the list will be picked
+    *   If none is registered, the config is invalid
+*   The config for the picked policy will be validated
+    *   If the validation fails, the config is invalid. We won’t move on to the
+        next entry.
+
+Action names are used as ID for the child policies. If an action name exists
+in the previous service config, the action’s balancer config will be sent as
+an update to the corresponding child policy (instead of creating a new child
+policy for it).
+
+NOTE: this LB policy is named `xds_routing` (instead of just routing) because
+we plan to make it deeply integrated with the xds resolver as part of
+RouteAction design.
+
+### Picking
+
+When making a pick, the first match wins.
+
+For example, with config
+
+```
+{prefix: "/MyService"}:             childPolicy-1
+{path:   "/MyService/MyMethod"}:    childPolicy-2
+```
+
+`ChildPolicy-1` will be picked for `/MyService/MyMethod`, even though the
+second route is an exact match. It is this way because Envoy always picks the
+first matched route. And with header matching, finding the exact match can be
+tricky.
+
+The pick will always be delegated to the picker from the matched child
+policy, even if it’s not in connectivity state READY (as opposed to
+weighted\_target will only pick the READY child policies).
+
+There will also be cases where for an RPC, a match is found, but the child
+policy hasn’t generated the picker (for example, the child policy is still
+initializing, and no picker has been created). When this happens, the RPCs
+will be always queued until a picker is updated.
```

Comment:
just to clarify the meaning: "the RPCs will always be queued at the matched child policy until a picker is created". Is that right?

Reply:
Not always waiting for the child picker. It can also be a new parent picker (meaning a new RDS response), which will select a different action.